Give react query

0xLienid opened the pull request:
Migrates the Give functionality to use react-query, which results in **MUCH** faster loading.

Also fixes some issues with incorrect metrics being displayed to the user.

0xJem committed:
Replace SvgIcon

0xJem committed:
Fix alignment of the education graphics

0xJem committed:
RecipientModal: migrate to Modal from component-library

0xJem committed:
Update jest snapshot

0xJem committed:
Fix rendering bug with modal on Firefox

0xJem committed:
Shift to component-library's Modal

0xJem committed:
Cleanup

0xJem committed:
Shifted ManageDonationModal to be less complicated

0xJem committed:
Migrate buttons

0xJem committed:
Adjust spacing for ArrowGraphic

0xJem committed:
Migrate ManageDonationModal to component-library. Fix build warnings.

0xLienid committed:
Update images, change button sizing

0xJem committed:
Correct implementation of modal onClose and back button

0xJem committed:
Migrate RedeemYieldModal to component-library

0xJem committed:
Migrate RedeemYield to use component-library

0xJem committed:
Remove redundant view

0xJem committed:
Migrate ProjectCard to component-library

0xJem committed:
Migrate YieldRecipients to component-library

0xJem committed:
Migrate GiveInfo

0xJem committed:
Partial migration of Give to component-library

0xJem committed:
Migrate Give to component-library

0xJem committed:
Migrate DepositRow to component-library

0xJem committed:
Partial migration of Causes Dashboard to component-library

0xJem committed:
Workaround for component-library issue

0xJem committed:
Merge branch 'develop' into tyche-causes

0xJem committed:
Merge branch 'tyche-causes' into tyche-component-library

0xJem committed:
Update jest snapshots

0xJem committed:
Merge branch 'tyche-component-library' of github.com:OlympusDAO/olympus-frontend into tyche-component-library

0xJem committed:
Fix failing test

0xJem committed:
Remove redundant CSS after component-library migration

0xLienid committed:
Reworked styling, updated short description copy, and added grants explainer

0xJem committed:
Partial fixes for spacing in project details

0xJem committed:
Cleanup spacing in project details

0xJem committed:
Fix positioning of edit button

0xJem committed:
Adjust spacing for countdown timer

0xJem committed:
Fix right-alignment of sOHM Yield Sent

0xJem committed:
Merge branch 'develop' into tyche-component-library

brightiron committed:
move this one component back to muipaper

0xJem committed:
Merge pull request #1481 from OlympusDAO/brightiron/muipaper-tyche

move this one component back to muipaper

0xJem committed:
Remove obsolete Typography components, update to component-library 1.7.0

0xJem committed:
Merge branch 'tyche-component-library' of github.com:OlympusDAO/olympus-frontend into tyche-component-library

0xJem committed:
Merge remote-tracking branch 'origin/develop' into tyche-grants

0xJem committed:
Define schema for milestones and shift GrantsDashboard to use it

0xJem committed:
WIP milestones in grants card

0xJem committed:
Merge remote-tracking branch 'origin/develop' into tyche-component-library

0xJem committed:
Fix failing test due to changes

0xJem committed:
Merge branch 'tyche-component-library' into tyche-grants

0xJem committed:
Compilation fix

0xJem committed:
Shift GrantCard to component-library

0xJem committed:
WIP milestone progress indicator

0xJem committed:
Return to grants page

0xJem committed:
WIP implementation of the milestone progress bar (using an external library)

0xJem committed:
Complete spacing and styling of milestone progress bar, plus milestone details.

0xJem committed:
Handle out of bonds index

0xJem committed:
Milestone amount

0xJem committed:
WIP migration of the grant data section to a responsive Grid

0xJem committed:
Bold

0xLienid committed:
Working on icon updates

0xJem committed:
Complete layout of grant data

0xJem committed:
Merge branch 'tyche-grants' of github.com:OlympusDAO/olympus-frontend into tyche-grants

0xJem committed:
Capitalisation

0xJem committed:
Fix displaying donor count

0xJem committed:
Remove extraneous log

0xJem committed:
Fix button position on edit donation card

0xJem committed:
Shift Grant screen to use Grid instead of hacky CSS

0xJem committed:
Remove extraneous classes

0xJem committed:
About section layout

0xJem committed:
Fix positioning on grants profile

0xJem committed:
Correct sizing of grant image

0xJem committed:
Shift more elements to Grid

0xJem committed:
Shift donations section to Grid

sam-potter committed:
Use common metric components

sam-potter committed:
Abstracted balances

sam-potter committed:
Combine switching/adding network to one hook

sam-potter committed:
Abstract switch network component

sam-potter committed:
Update test snapshot

0xJem committed:
Shift to using the topLeft prop of the Paper component and Grid

0xJem committed:
Fix sizing of grant image in medium breakpoint

0xJem committed:
Correct sizing of milestone details

0xJem committed:
Add new icons (though they have theme problems)

0xJem committed:
Correct sizing of SVG icons

0xJem committed:
Merge remote-tracking branch 'origin/develop' into tyche-grants

0xJem committed:
Update failing snapshot test

0xJem committed:
Remove redundant code

sam-potter committed:
Convert mutations

sam-potter committed:
Update test snapshots

sam-potter committed:
Merge branch 'develop' into react-query/wrap-view

0xJem committed:
Align grant profile with design 2.0.2

0xJem committed:
Add the ability to hide a grant from display

0xJem committed:
Update grants milestone amounts

0xJem committed:
Fix rendering of the donors icon with themes. This required explicitly adding a fill property, and defining the stroke property as currentcolor

0xJem committed:
Workaround for SVG fills

0xJem committed:
Add remaining 2 SVG icons for Grants (with stroke/fill fixes)

0xJem committed:
Fix wrapping of donors

0xJem committed:
Remove redundant grant

0xJem committed:
Didn't mean to enable that

sam-potter committed:
Abstract WalletConnectedGuard

sam-potter committed:
Convert cross chain migration

sam-potter committed:
Cleanup

0xJem committed:
Correct layout on md and lg breakpoints

sam-potter committed:
Update test snapshots

sam-potter committed:
Merge branch 'develop' into react-query/wrap-view

sam-potter committed:
Stop making unecessary requests to testnets

sam-potter committed:
Add react query error handler

0xJem committed:
Fix SVG rendering of fills

0xJem committed:
SVG updates
- Add new/updated icons
- Remove redundant icons
- Fix SVG background/fill issues

These SVG icons will be moved into component-library shortly.

0xJem committed:
Update EducationCard with latest SVG icons

0xJem committed:
Merge remote-tracking branch 'origin/develop' into tyche-grants

0xJem committed:
Hide project stats for grants

0xJem committed:
Merge remote-tracking branch 'origin/HEAD' into tyche-grants

0xJem committed:
Remove extraneous log

0xJem committed:
Update to component-library 1.7.1

0xJem committed:
Shift icons in GrantCard to component-library

0xJem committed:
Migrate ProjectCard to component-library icons

0xJem committed:
Shift EducationCard to use component-library icons

0xJem committed:
Remove redundant SVG files

0xJem committed:
Add feature flag to disable Give Grants

0xJem committed:
Misc design fixes

0xJem committed:
Update snapshots

0xJem committed:
Disable individual grant projects until they are approved. Also display helpful text if no grants are active.

0xJem committed:
Redundant Grid component

0xJem committed:
Remove redundant CSS

0xJem committed:
Shift CauseCard to use Grid component

0xJem committed:
WIP refactoring of ProjectCard to use Grid component

0xJem committed:
Remove redundant className

0xJem committed:
Remove redundancies

0xJem committed:
Revert temporary change to user donating flag

0xJem committed:
Remove redundant CSS

0xJem committed:
Remove redundant CSS

0xJem committed:
Update test snapshot with recent changes

0xJem committed:
Remove redundant CSS

0xJem committed:
Simplify the image used in Project and Grant screens

0xJem committed:
Convert amount screen of RecipientModal to use the Grid component

0xJem committed:
WIP conversion of RecipientModal to Grid component

0xJem committed:
Fix breakpoints on confirmation screen.

0xJem committed:
Align RecipientModal with reference design

0xJem committed:
Migrate GiveInfo. Remove redundant CSS.

0xJem committed:
Update snapshot

0xJem committed:
Shift CausesDashboard to use MUI components

0xJem committed:
Cleanup Give component

0xJem committed:
Update snapshot

0xJem committed:
Migrate CausesDashboard and GrantsDashboard

0xJem committed:
Clean up redundant CSS (LOTS of it)

0xJem committed:
Fix breakpoints for project and grant card

0xJem committed:
Responsive version of no donations message

0xJem committed:
Migrate manage donation screen

0xJem committed:
Migrate ManageDonationModal

0xJem committed:
Shift to use GiveBox

0xJem committed:
Convert RedeemYield to use Grid. Remove redundant CSS.

0xJem committed:
Use the DataRow component

0xJem committed:
Remove redundant component

0xJem committed:
Adjust breakpoint

0xJem committed:
Accidental commit

0xJem committed:
Migrate RedeemYieldModal to use Grid. Cleanup CSS.

0xJem committed:
Revert feature flag for give grants (since we will display an informative message instead)

sam-potter committed:
Merge branch 'develop' into react-query/wrap-view

sam-potter committed:
Update import

sam-potter committed:
Update test snapshots

sam-potter committed:
Move file

sam-potter committed:
Add react-query devtools

sam-potter committed:
Add error handler

0xJem committed:
Merge branch 'develop' into tyche-grants

# Conflicts:
#	.env.example
#	.env.test
#	src/locales/translations

0xJem committed:
Compilation fixes due to merge

0xJem committed:
Rebase fix

0xJem committed:
Layout fixes. Remove custom styling where possible.

0xJem committed:
Display message if there is no milestone information

0xJem committed:
Clean up EducationCard: shift to Grid components, remove redundant graphics, rename components.

0xJem committed:
Remove redundant SVG fill

sam-potter committed:
Add support for migrating specific amounts

sam-potter committed:
Refactor

sam-potter committed:
Remove unused migrate functions

sam-potter committed:
Fix translations

sam-potter committed:
Merge branch 'develop' into react-query/wrap-view

sam-potter committed:
Use testnets

sam-potter committed:
Prefer global query error handler

sam-potter committed:
Fix incorrect boolean

sam-potter committed:
Use testnets

sam-potter committed:
Abstract ReactQueryProvider

0xJem committed:
Remove hard-coded temporary value

0xJem committed:
Center-align the right-facing arrow

0xJem committed:
Backport text alignment

0xJem committed:
Update snapshot

0xJem committed:
Merge branch 'develop' into tyche-grants

0xJem committed:
Merge branch 'tyche-grants' into give-project-card

sam-potter committed:
Merge branch 'develop' into react-query/wrap-view

sam-potter committed:
Use new DecimalBigNumber helper

0xJem committed:
Merge pull request #1545 from OlympusDAO/give-project-card

Migrate Give screens to Grid component

0xJem committed:
Fix Paper width. Remove extraneous padding around causes/grant dashboards.

0xJem committed:
Apply fullWidth to Paper

0xJem committed:
Merge branch 'develop' into tyche-grants

# Conflicts:
#	src/views/Give/RecipientModal.tsx

0xJem committed:
Better handling of the main Give container at different breakpoints

0xJem committed:
Update snapshots

0xLienid committed:
Adding addresses

0xLienid committed:
merge off of react-query/wrap-view

0xLienid committed:
Writing hooks

0xLienid committed:
Update token guard for Give modal, put in file for getting Give info

0xLienid committed:
Update to using hooks for contract interactions

0xLienid committed:
Adding new query hooks to pull data. Only works for mainnet right now

0xLienid committed:
useDonationInfo hook

0xLienid committed:
Hooks update on network change

0xLienid committed:
Cleaning up files

0xLienid committed:
Fixing components so data refreshes appropriately

0xLienid committed:
Merge with develop

0xLienid committed:
Re-add test snapshots

0xLienid committed:
Fix merge issue in useProtocolMetrics

0xJem commented on src/components/GiveProject/GrantCard.tsx:
```diff
@@ -318,37 +268,7 @@ export default function GrantCard({ grant, mode }: GrantDetailsProps) {
 
-    // If on Rinkeby and using Mock Sohm, use changeMockGive async thunk
-    // Else use standard call
-    if (networkId === NetworkId.TESTNET_RINKEBY && Environment.isMockSohmEnabled(location.search)) {
```
I don't see any use of the mock sOHM contract in the new implementation. Do we really want to remove it?

0xLienid replied:
No, we probably still do want it, just was planning on waiting for the new YieldDirector contract to finish it's audit and deploy a version of that with mock sOHM rather than building all the hooks for mock sOHM, then in 2 days having to convert double the work to the new contract

0xJem commented on src/views/Give/hooks/useGive.ts:
```diff
@@ -0,0 +1,46 @@
+    async ({ amount: amount_, recipient: recipient_ }) => {
+      if (!contract) throw new Error(t`Please switch to the Ethereum network to donate yield`);
+
+      const transaction = await contract.deposit(ethers.utils.parseUnits(amount_, "gwei"), recipient_);
```
As I understand it, this replaces the functions in GiveThunk. Is that correct?

If so:
- Remove the GiveThunk functions/file
- Add the Google Analytics event call

0xJem commented on src/views/Give/hooks/useGive.ts:
```diff
@@ -0,0 +1,46 @@
+
+  return useMutation<ContractReceipt, Error, GiveData>(
+    async ({ amount: amount_, recipient: recipient_ }) => {
+      if (!contract) throw new Error(t`Please switch to the Ethereum network to donate yield`);
```
More accurate would be: "Give is not supported on this network. Please switch to a supported network, such as Ethereum mainnet."

0xJem commented on src/components/GiveProject/ProjectCard.tsx:
```diff
@@ -211,12 +163,12 @@ export default function ProjectCard({ project, mode }: ProjectDetailsProps) {
     const totalDonatedNumber = new BigNumber(totalDonated);
 
-    return totalDonatedNumber.div(depositGoal).multipliedBy(100).toFixed(2);
+    return totalDonatedNumber.div(depositGoal).multipliedBy(100).toNumber().toFixed(2);
```
Why? AFAIK `multipliedBy` returns a BigNumber, and we call `toFixed` on that.

0xLienid replied:
yeah think I messed these up when trying different bignumber implementations. fixed now

0xJem commented on src/components/GiveProject/ProjectCard.tsx:
```diff
@@ -254,7 +206,11 @@ export default function ProjectCard({ project, mode }: ProjectDetailsProps) {
-    const formattedTotalDonated = new BigNumber(parseFloat(totalDonated).toFixed(2)).toFormat();
+    const formattedTotalDonated = !totalDonated
+      ? new BigNumber("0")
+      : new BigNumber(totalDonated).toNumber().toFixed(2);
```
You should not need to convert the BigNumber to a number in order to use toFixed().

0xJem commented on src/components/GiveProject/ProjectCard.tsx:
```diff
@@ -254,7 +206,11 @@ export default function ProjectCard({ project, mode }: ProjectDetailsProps) {
+      ? new BigNumber("0")
+      : new BigNumber(totalDonated).toNumber().toFixed(2);
+
+    if (depositGoal === 0) return <></>;
```
Duplicated

0xJem commented on src/components/TokenAllowanceGuard/TokenAllowanceGuard.tsx:
```diff
@@ -81,3 +81,48 @@ export const TokenAllowanceGuard: React.FC<{
   return <>{props.children}</>;
 };
+
+export const GiveTokenAllowanceGuard: React.FC<{
```
I would rename the filename to match the component nane.

0xJem commented on src/components/TokenAllowanceGuard/TokenAllowanceGuard.tsx:
```diff
@@ -81,3 +81,48 @@ export const TokenAllowanceGuard: React.FC<{
+
+  if (allowance.eq(0))
+    return (
+      <Grid container className={classes.inputRow} style={{ display: "flex", flexDirection: "column" }}>
```
Use the native Grid attributes instead of custom styles.

0xJem replied:
Fixed.

0xJem commented on src/components/TokenAllowanceGuard/TokenAllowanceGuard.tsx:
```diff
@@ -81,3 +81,48 @@ export const TokenAllowanceGuard: React.FC<{
+  const approveMutation = useApproveToken(props.tokenAddressMap, props.spenderAddressMap);
+  const { data: allowance } = useContractAllowance(props.tokenAddressMap, props.spenderAddressMap);
+
+  if (!allowance)
```
Document what this is doing and why

0xJem commented on src/hooks/useGiveInfo.ts:
```diff
@@ -0,0 +1,245 @@
+        try {
+          yieldSent = await contract.donatedTo(address, allDeposits[0][i]);
+        } catch (e: unknown) {
+          console.log(
```
console.error

0xJem commented on src/hooks/useGiveInfo.ts:
```diff
@@ -0,0 +1,245 @@
+      for (let i = 0; i < allDeposits[0].length; i++) {
+        if (allDeposits[1][i].eq(0)) continue;
+
+        const firstDonationDate = await GetDonationDate({
```
Can you extract this out of the `for` loop? It's inefficient and slow.

0xJem commented on src/hooks/useGiveInfo.ts:
```diff
@@ -0,0 +1,245 @@
+        const firstDonationDate = await GetDonationDate({
+          address: address,
+          recipient: allDeposits[0][i],
+          networkID: networks.MAINNET,
```
Why is this locked to mainnet? What if it's being run on Rinkeby?

0xJem replied:
Is there a reason for this being mainnet?

0xJem commented on src/hooks/useGiveInfo.ts:
```diff
@@ -0,0 +1,245 @@
+
+        let yieldSent: BigNumber = BigNumber.from("0");
+        try {
+          yieldSent = await contract.donatedTo(address, allDeposits[0][i]);
```
Same with this. Extract out of the loop.

0xJem commented on src/hooks/useGiveInfo.ts:
```diff
@@ -0,0 +1,245 @@
+    async () => {
+      queryAssertion([address, networkId], redeemableBalanceQueryKey(address, networkId));
+
+      if (!contract) throw new Error(t`Please switch to the Ethereum network`);
```
Same as before. A more descriptive error.

0xJem commented on src/hooks/useGiveInfo.ts:
```diff
@@ -0,0 +1,245 @@
+        const event = events[i];
+
+        if (event.topics[2] === zeroPadAddress.toLowerCase()) {
+          const donorActiveDonations: [string[], BigNumber[]] = await contract.getAllDeposits(
```
Pull out of for loop

0xJem replied:
Have you faced any difficulties with this?

0xJem replied:
Never mind. I don't think this is necessary.

0xJem commented on src/views/Give/DepositRow.tsx:
```diff
@@ -32,16 +28,17 @@ interface DepositRowProps {
   const [isManageModalOpen, setIsManageModalOpen] = useState(false);
   const theme = useTheme();
-  const isSmallScreen = useMediaQuery(theme.breakpoints.down("xs"));
+  const isSmallScreen = useMediaQuery(theme.breakpoints.down("sm"));
```
Please restore unless there's a good reason for this.

0xJem replied:
Fixed.

0xJem commented on src/views/Give/DepositRow.tsx:
```diff
@@ -66,38 +63,13 @@ export const DepositTableRow = ({ depositObject }: DepositRowProps) => {
-          eventSource,
-        }),
-      );
+      const subtractionAmount = depositAmountDiff.multipliedBy(new BigNumber("-1"));
```
`depositAmountDiff` is being tested for being positive above, which implies that it can be negative. Here you're multiplying a negative number by -1, which would result in a positive number. Whoops.

0xJem commented on src/views/Give/DepositRow.tsx:
```diff
@@ -66,38 +63,13 @@ export const DepositTableRow = ({ depositObject }: DepositRowProps) => {
-          eventSource,
-        }),
-      );
+    if (depositAmountDiff.eq(new BigNumber("0"))) return;
```
This code is duplicated from ProjectCard/GrantCard, IIRC. Is there some way to consolidate these callback functions?

0xJem commented on src/views/Give/ManageDonationModal.tsx:
```diff
@@ -202,7 +176,7 @@ export function ManageDonationModal({
     if (!address) return false;
     if (hasPendingGiveTxn(pendingTransactions)) return false;
-    if (getDepositAmountDiff().isEqualTo(0)) return false;
+    if (getDepositAmountDiff().isEqualTo(new BigNumber("0"))) return false;
```
Why? BigNumber(.js) supports this.

0xJem commented on src/views/Give/ManageDonationModal.tsx:
```diff
@@ -215,7 +189,15 @@ export function ManageDonationModal({
+    return sohmBalance ? new BigNumber(sohmBalance.toAccurateString()) : new BigNumber(0);
+  };
+
+  const getTotalDebt = (): string => {
```
Why not return a BigNumber?

0xJem commented on src/views/Give/ManageDonationModal.tsx:
```diff
@@ -215,7 +189,15 @@ export function ManageDonationModal({
+    return totalDebt ? totalDebt : "";
+  };
+
+  const getTotalDonated = (): string => {
```
Why not return a BigNumber?

0xJem commented on src/views/Give/ManageDonationModal.tsx:
```diff
@@ -424,7 +406,7 @@ export function ManageDonationModal({
           <Box>
             <Typography variant="h5" align="center">
-              {project ? parseFloat(totalDebt).toFixed(2) : "N/A"}
+              {project ? parseFloat(getTotalDebt()).toFixed(2) : "N/A"}
```
Use a BigNumber and avoid `parseFloat`. Then you can use BigNumber's `toFixed` method.

0xJem commented on src/views/Give/RecipientModal.tsx:
```diff
@@ -351,118 +330,95 @@ export function RecipientModal({ isModalOpen, eventSource, callbackFunc, cancelF
+                value={getDepositAmount().eq(0) ? null : getDepositAmount()}
+                helperText={
+                  isDepositAmountValid
+                    ? `${t`Your current Staked Balance is`} ${getSOhmBalance().toNumber().toFixed(2)} sOHM`
```
No need to convert toNumber(), unless I'm missing something?

0xJem commented on src/views/Give/RedeemYield.tsx:
```diff
@@ -75,10 +68,10 @@ export default function RedeemYield() {
    */
   const getTrimmedBigNumber = (number: BigNumber) => {
-    return number.decimalPlaces(4).toString();
+    return number.toNumber().toFixed(4).toString();
```
toNumber()

0xJem commented on src/views/Give/RedeemYield.tsx:
```diff
@@ -97,21 +90,10 @@ export default function RedeemYield() {
+    if (!totalDonated) return 0;
 
-    return 0;
+    return parseFloat(totalDonated);
```
Return a BigNumber

0xJem commented on src/views/Give/YieldRecipients.tsx:
```diff
@@ -58,6 +50,7 @@ export default function YieldRecipients({ changeView }: RecipientModalProps) {
     );
   }
 
+  // TODO extract the table and styles into a common component
```
Not needed

0xJem commented on src/views/Give/YieldRecipients.tsx:
```diff
@@ -20,17 +17,12 @@ type RecipientModalProps = {
-  const { networkId } = useWeb3Context();
   const theme = useTheme();
-  const isSmallScreen = useMediaQuery(theme.breakpoints.down("xs"));
+  const isSmallScreen = useMediaQuery(theme.breakpoints.down("sm"));
```
Revert

0xJem commented on src/views/Give/hooks/useEditGive.ts:
```diff
@@ -0,0 +1,80 @@
+        const promises = keysToRefetch.map(key => client.refetchQueries(key, { active: true }));
+        await Promise.all(promises);
+
+        dispatch(createInfoToast(t`Successfully deposited sOHM`));
```
`Successfully increased sOHM deposit`

0xJem commented on src/views/Give/hooks/useEditGive.ts:
```diff
@@ -0,0 +1,80 @@
+        const promises = keysToRefetch.map(key => client.refetchQueries(key, { active: true }));
+        await Promise.all(promises);
+
+        dispatch(createInfoToast(t`Successfully deposited sOHM`));
```
`Successfully decreased sOHM deposit`

0xJem reviewed (changes requested):
Overall, this seems like a good move.

A few overarching comments:
- When there's a `BigNumber`, I see you're calling `toNumber().toFixed(2)`, where previously `toFixed(2)` was being called. I don't see a good reason for that, and it's probably more imprecise.
- There are some instances where `number` is being passed around, and `BigNumber` should be used instead.
- We need to standardise on `DecimalBigNumber` at some point, though it is probably non-trivial and out of scope for this PR. 
- More function and code comments please. 
- You also have test failures

0xJem commented:
Entering "1.0" into the ManageDonationModal doesn't work. It deletes the ".0". This would prevent the user from writing "1.01", for example.

RecipientModal has the same problem. 

EDIT: Turns out this is a problem on prod and staging. Not introduced in this PR.

0xJem commented:
RecipientModal:
- The "Donate Yield" button never shows a pending status now, after approving the MM transaction
- The interface doesn't update after the donation transaction has been approved (but it does after a reload)  

ManageDonationModal:
- "Edit Donation" does not show pending status
- Interface doesn't update post-transaction  

YieldRecipients:
- It still seems really slow to load, even on subsequent refreshes

Donate, edit, withdraw all seem to work.

0xLienid committed:
Documentation

0xLienid commented:
interface updating data after transaction works now. working on pending status

0xLienid committed:
BigNumber, pending transactions, removal of GiveThunk

0xLienid committed:
Fix merge issue on tests

0xLienid committed:
Eliminate excess loading flags for faster data display

0xLienid committed:
Button states, GA tracking

0xJem commented on src/components/GiveProject/GrantCard.tsx:
```diff
@@ -54,22 +54,32 @@ export default function GrantCard({ grant, mode }: GrantDetailsProps) {
 
+  useEffect(() => {
+    setIsUserDonating(false);
+    setDonationId(0);
```
If this is 0-indexed, it could be problematic. How about defining a named constant with a value of -1?

0xJem commented on src/components/GiveProject/ProjectCard.tsx:
```diff
@@ -206,9 +219,7 @@ export default function ProjectCard({ project, mode }: ProjectDetailsProps) {
-    const formattedTotalDonated = !totalDonated
-      ? new BigNumber("0")
-      : new BigNumber(totalDonated).toNumber().toFixed(2);
+    const formattedTotalDonated = !totalDonated ? new BigNumber("0") : new BigNumber(totalDonated).toFixed(2);
```
If you set the type of `formattedTotalDonated` to `string`, you'll notice that one path of the conditional sets it to BigNumber(.js) and the other a string. Whoops.

0xJem commented on src/hooks/useGiveInfo.ts:
```diff
@@ -35,21 +50,31 @@ export const useDonationInfo = () => {
 
-      if (!contract) return donationInfo;
+      // If there's no contract (i.e. on a non-ETH network), throw error
+      if (!contract)
```
How does the frontend handle throwing this error? Please test that. Does it cause the frontend to crash? Probably a bad thing. Perhaps it should call `dispatch(error("etc..."))`? But the calling function should handle this accordingly.

0xLienid replied:
It just reports it as an error in console like this. what do you think?: 
<img width="1710" alt="image" src="https://user-images.githubusercontent.com/92545857/162282949-55dbf70c-874a-4c14-bcdf-8f1cf32d2edf.png">

0xJem replied:
I think it'll also show an error message in the interface (based on my experience with it in OlyZaps), which is fine.

0xJem commented on src/views/Give/Give.tsx:
```diff
@@ -23,6 +23,13 @@ function a11yProps(index: number) {
   };
 }
 
+export const isSupportedChain = (chainID: NetworkId): boolean => {
```
It doesn't make sense for this helper function to be in a view. Consider shifting it to another file.

0xJem commented on src/views/Give/hooks/useEditGive.ts:
```diff
@@ -6,74 +6,126 @@ import { GIVE_ADDRESSES, SOHM_ADDRESSES } from "src/constants/addresses";
+      // Create transaction to deposit passed amount to the passed recipient
       const transaction = await contract.deposit(ethers.utils.parseUnits(amount_, "gwei"), recipient_);
+
+      const uaData: IUAData = {
```
I pushed some changes a few weeks back that logs a GA event before and after the transaction (as we were missing events and wanted to find out why). Can you ensure this is replicated here?

https://github.com/OlympusDAO/olympus-frontend/pull/1574/files

0xJem reviewed (approved):
A few remaining changes, plus the extraction of async/await calls out of for loops (unless you disagree).

0xJem reviewed (changes requested):
> A few remaining changes, plus the extraction of async/await calls out of for loops (unless you disagree).

Oops.

0xJem commented:
> RecipientModal:
> 
> * The "Donate Yield" button never shows a pending status now, after approving the MM transaction
> * The interface doesn't update after the donation transaction has been approved (but it does after a reload)
> 
> ManageDonationModal:
> 
> * "Edit Donation" does not show pending status
> * Interface doesn't update post-transaction
> 
> YieldRecipients:
> 
> * It still seems really slow to load, even on subsequent refreshes
> 
> Donate, edit, withdraw all seem to work.

These seem to be fixed.

Also, the data loading is remarkably faster than before. Well done!

https://user-images.githubusercontent.com/91760036/161223109-bcee26ff-d41c-4885-9b13-5edf24d5371d.mp4

(Your changes also seem to fix a bug with "sOHM Deposited" showing 0.)

0xLienid committed:
Extracting helpers, adding a 'before' GA event, minor tweaks

0xJem commented on src/components/GiveProject/GrantCard.tsx:
```diff
@@ -84,41 +78,9 @@ export default function GrantCard({ grant, mode }: GrantDetailsProps) {
-  }, [connected, networkId, isGiveModalOpen]);
+    setIsUserDonating(false);
+    setDonationId(NO_DONATION);
+  }, [networkId]);
```
If the networkId changes, don't we also need to invalidate the fetched data? Unless you've included the networkId as part of the react-query key.

0xJem replied:
(You have, so ignore this.)

0xJem commented on src/components/GiveProject/GrantCard.tsx:
```diff
@@ -242,7 +207,7 @@ export default function GrantCard({ grant, mode }: GrantDetailsProps) {
                   </Grid>
                   <Grid item className="metric">
-                    {donorCountIsLoading ? <Skeleton className="skeleton-inline" /> : donorCount}
+                    {!donorCount ? <Skeleton className="skeleton-inline" /> : donorCount}
```
`useDonationInfo().isLoading` is what you should be accessing here.

If there are no donors to a project, `!donorCount` will return `true` and you'll show the skeleton component.

0xJem commented on src/components/GiveProject/GrantCard.tsx:
```diff
@@ -610,7 +485,7 @@ export default function GrantCard({ grant, mode }: GrantDetailsProps) {
                             <Grid item>
                               <Typography className="metric">
-                                {donationInfo[donationId]
+                                {donationId != NO_DONATION && donationInfo[donationId]
```
You could probably simplify this and use something like:

```ts
const currentDonation = useMemo(() => {
   if (donationId == NO_DONATION) return null;

   return donationInfo[donationId];
}, [donationInfo, donationId]);
```
(please excuse typos)

then this (and subsequent) conditionals change to:
```ts
currentDonation ? parseFloat(currentDonation.yieldDonated).toFixed(2) : "0"
```

0xJem commented:
I would suggest that you rebase off #1562 , since that has a bunch of formatting changes and they may conflict with what you have

0xLienid commented:
Re: "why mainnet", I believe that useTestableNetworks().MAINNET just means mainnet in the sense of Ethereum vs other chains. It looks like it returns either Ethereum Mainnet or Rinkeby depending on what you are connected to, so I just used that.

`  return {
    MAINNET: getTestnet(NetworkId.MAINNET, NetworkId.TESTNET_RINKEBY, networkId),
    AVALANCHE: getTestnet(NetworkId.AVALANCHE, NetworkId.AVALANCHE_TESTNET, networkId),
    ARBITRUM: getTestnet(NetworkId.ARBITRUM, NetworkId.ARBITRUM_TESTNET, networkId),
    POLYGON: getTestnet(NetworkId.POLYGON, NetworkId.POLYGON_TESTNET, networkId),
    FANTOM: getTestnet(NetworkId.FANTOM, NetworkId.FANTOM_TESTNET, networkId),
  };`

0xLienid commented:
I haven't noticed any issues with that call. why?

0xLienid commented:
and ok, will rebase

0xLienid committed:
Review tweaks

0xLienid commented:
oh boy this merge is ugly and is going to take a while

0xJem commented:
@0xLienid the other non-trivial change that I see remaining if that there are a few `await` calls (I assume making queries on the chain) within for loops, which is going to be slow, expensive and is generally not a good idea. Can you refactor those?

0xJem committed:
Compilation fix

0xJem committed:
Fix variable dependencies

0xJem committed:
Merge branch 'develop' into give-react-query

# Conflicts:
#	src/components/GiveProject/GrantCard.tsx
#	src/components/GiveProject/ProjectCard.tsx
#	src/locales/translations
#	src/slices/GiveThunk.ts
#	src/views/Give/CausesDashboard.tsx
#	src/views/Give/DepositRow.tsx
#	src/views/Give/GrantsDashboard.tsx
#	src/views/Give/ManageDonationModal.tsx
#	src/views/Give/RecipientModal.tsx
#	src/views/Give/RedeemYield.tsx

0xJem committed:
Remove redundant variables, simplify

0xJem committed:
Fix infinite loop

0xJem commented:
What is the difference between what is returned by `useTotalDonated().data` and useRecipientInfo().data.totalDebt`? See ManageDonationModal#386. We were using `totalDonated`, but we should have been using `totalDebt`. Confusing. Consider renaming.

0xJem committed:
Fix percentage calculation

0xJem commented:
One issue I've noticed: causes dashboard -> view profile (when you have already donated) does not load the "user donations" section. But if you refresh, it does.

0xJem committed:
Cleanup of variables

0xJem committed:
Cleanup variables

0xJem committed:
Remove access to state and old cruft

0xJem commented:
I've completed the merge of `develop` and performed some additional cleanup and fixes (e.g. totalDebt instead of totalDonated)

What I think is left:
- [x] Clarity on debt vs donated: https://github.com/OlympusDAO/olympus-frontend/pull/1618#issuecomment-1094719874 -- done
- [x] Loading user donations: https://github.com/OlympusDAO/olympus-frontend/pull/1618#issuecomment-1094722705 and https://www.notion.so/olympusdao/User-Donations-do-not-appear-when-navigating-to-profile-7e76034559d54a58a41a3c9d67c0db51 (I tried fixing this... couldn't) -- done
- [x] Migrate RedeemYieldModal to use react-query
- [x] Remove redundant state-related code, e.g. `DonationInfoState` and from AccountSlice
- [x] Update RedeemYield tests to use react-query (there are incorrect values compared to the snapshot)

0xJem committed:
Update snapshots (removed skeleton component)

0xJem committed:
Fix incorrect variable used

0xJem committed:
Fix bug with skeleton component on project cards

0xJem committed:
Cleanup debt/deposit confusion

0xJem committed:
Rename query with ambiguous name

0xJem committed:
Renaming of ambiguous method

0xJem committed:
Fix bug that resulted a user's donation information not loading

0xJem committed:
Remove redundant items in AccountSlice. Mark other code to be removed.

0xJem committed:
Remove redundant test code

0xJem committed:
Minor console fix

0xLienid committed:
Move awaits out of loops. useDonorNumbers returns a sum of mainnet and testnet though...

0xLienid committed:
Remove console logs, add redeem hook

0xLienid committed:
Update Give test snapshots, and mock returns for redeem yield hooks

0xLienid committed:
Clean up Donation Info hook

0xLienid committed:
Remove redundant files and functions, test updates, modal state updates

0xJem commented on src/hooks/useGiveInfo.ts:
```diff
@@ -61,42 +61,59 @@ export const useDonationInfo = () => {
+        firstDonationData = await Promise.all(firstDonationDatePromises);
+        yieldSentData = await Promise.all(yieldSentPromises);
+      } catch (e: unknown) {
+        console.info(
```
You've already caught this error in line 89, is there a need to catch it again?

0xJem commented on src/views/Give/RedeemYieldModal.tsx:
```diff
@@ -98,14 +97,7 @@ export function RedeemYieldModal({ isModalOpen, callbackFunc, cancelFunc, redeem
-                    format: true,
-                  })} sOHM`,
-                )}
+                {isMutationLoading ? t`Redeeming ` : t`Redeem`} sOHM
```
Need to add the number of sOHM being redeemed

0xJem replied:
Fixed

0xJem commented on src/views/Give/__tests__/__snapshots__/RedeemYield.unit.test.jsx.snap:
```diff
@@ -181,7 +181,9 @@ exports[`Redeem Yield should show extra content if project wallet 1`] = `
         data-testid="redeemable-balance"
       >
-        100
+        <span
```
I didn't commit this change to the snapshot, as I think the snapshot should show this value. It shows a mis-configuration in the test.

0xJem replied:
Fixed

0xJem commented on src/views/Give/utils/googleAnalytics.ts:
```diff
@@ -10,11 +9,31 @@ export interface IUAData {
   trackGAEvent({
     category: "Olympus Give",
     action: eventAction ? eventAction : uaData.type ? uaData.type : "unknown",
-    label: getGiveProjectName(uaData.recipient) ?? "unknown",
```
I don't see why we should remove the usage of `getGiveProjectName` here

0xJem committed:
Restore redeem amount to modal

0xJem committed:
Cleanup styling in GiveTokenAllowanceGuard. Cleaner use of react-query hook.

0xJem committed:
Restore breakpoint

0xJem committed:
Restore project/grant name to Google Analytics events

0xJem committed:
Renaming artifact

0xJem committed:
Update trivial/styling changes in snapshot

0xJem committed:
Revert snapshot changes

0xJem committed:
Remove redundant code

0xJem committed:
Test fixes for RedeemYield

- Update snapshots with minor styling info
- Update snapshots with metrics that were overwritten
- Shift RedeemYield tests to mock react-query

0xJem committed:
Add test case for 0 redeemable balance

0xJem committed:
Constantly shifting snapshots

0xJem committed:
Merge branch 'develop' into give-react-query

0xJem reviewed (approved):
I'm happy with the current status. All tests are passing. @0xLienid want to push this for review?

0xLienid committed:
Removing redeeming and mockRedeeming from Account reducer

0xJem committed:
Remove redundant code
